### PR TITLE
Add telemetry_test

### PR DESCRIPTION
**Pull request by whatyouhide**
Hey friends!

I find myself doing this constantly. I think it would be good to have something built-in just for the sake of inviting people to test their telemetry events 😄 

I threw a PR together since I thought it'd be easier to talk about something concrete, but I’m not attached to API or implementation here.

Thoughts?

### CODE REVIEW

**Comment by arkgil** [comment]
Thanks, I like that!

File: src/telemetry_test.erl [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+%% Testing that the correct Telemetry events are emitted with the
+%% right measurements and metadata is essential for library authors.
+%% It helps to maintain stable APIs and avoid accidental changes
+%% to events.
+%% @end
+%%%-------------------------------------------------------------------
+
+-module(telemetry_test).
+
+-export([attach_message_handlers/3]).
```
Comment (arkgil) [inline src/telemetry_test.erl:13]:
```suggestion
-export([attach_event_handlers/3]).
```

Comment (arkgil) [reply src/telemetry_test.erl:13]: WDYT?

File: src/telemetry_test.erl [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+%% ```
+%% {Event, Ref, Measurements, Metadata}
+%% '''
+%%
+%% You can detach this event with {link telemetry:detach/1}.
+-spec attach_message_handlers(HandlerID, DestinationPID, Events) -> reference() when
+    HandlerID :: telemetry:handler_id(),
+    DestinationPID :: pid(),
+    Events :: [telemetry:event_name(), ...].
+attach_message_handlers(HandlerID, DestPID, Events) when is_pid(DestPID) and is_list(Events) ->
```
Comment (arkgil) [inline src/telemetry_test.erl:32]: How about we autogenerate the handler ID and return if from this function? I guess in most cases people wouldn't care about having specific handler ID.

File: src/telemetry_test.erl (hunk 1 above)
Comment (whatyouhide) [reply src/telemetry_test.erl:13]: Yes, sounds good!

File: src/telemetry_test.erl (hunk 2 above)
Comment (whatyouhide) [reply src/telemetry_test.erl]: Yes, fair enough. Done.

**Comment by whatyouhide** [comment]
@arkgil thanks for the review! I addressed your comments, waiting on CI now.

File: src/telemetry_test.erl [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+%% {Event, Ref, Measurements, Metadata}
+%% '''
+%%
+%% You can detach this event with {link telemetry:detach/1}.
+-spec attach_event_handlers(DestinationPID, Events) -> {telemetry:handler_id(), reference()} when
+    DestinationPID :: pid(),
+    Events :: [telemetry:event_name(), ...].
+attach_event_handlers(DestPID, Events) when is_pid(DestPID) and is_list(Events) ->
+    Ref = make_ref(),
+    HandlerId = crypto:strong_rand_bytes(16),
```
Comment (josevalim) [inline src/telemetry_test.erl:33]: The `Ref` itself should be good enough as ID, no?

File: src/telemetry_test.erl [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+-export([attach_event_handlers/2]).
+
+%% @doc Attaches a "message" handler to the given events.
+%%
+%% `handler_id` must be unique and is used as the handler ID.
+%%
+%% The attached handler sends a message to `destination_pid` every time it handles one of the
+%% events in `events`. The function returns a reference that you can use to make sure that
+%% messages come from this handler. The shape of messages sent to `destination_pid` is:
```
Comment (wojtekmach) [inline src/telemetry_test.erl:21]:
```suggestion
%% @doc Attaches a "message" handler to the given events.
%%
%% The attached handler sends a message to `DestinationPID' every time it handles one of the
%% events in `Events'. The function returns a reference that you can use to make sure that
%% messages come from this handler. The shape of messages sent to `DestinationPID' is:
```

File: src/telemetry_test.erl (hunk 3 above)
Comment (whatyouhide) [reply src/telemetry_test.erl:33]: Yes, I think that's a good idea 👍. Updated.

File: src/telemetry_test.erl (hunk 4 above)
Comment (whatyouhide) [reply src/telemetry_test.erl:21]: Ah, thanks @wojtekmach. I cleaned up the docs 👍

**Comment by whatyouhide** [comment]
@arkgil anything I can do to push this over the finish line? 🙃

**Comment by arkgil** [comment]
@whatyouhide apologies, I was on holiday. Would you mind adding `version-type: strict` below this line https://github.com/beam-telemetry/telemetry/blob/dc1cf9e0d71739485272fb38cec6e96b5e6bdfe9/.github/workflows/test.yml#L27 to see if this fixes the CI for other versions? Unless I'm missing something and it's not the reason for failure.

**Comment by whatyouhide** [comment]
@arkgil done 👍

**Comment by arkgil** [comment]
@whatyouhide I fixed the CI in https://github.com/beam-telemetry/telemetry/pull/120, after that's merged can you please rebase your PR on top of main? Thanks!

**Comment by josevalim** [comment]
Can we include examples of how a test would look like in Erlang and Elixir? Something like attach, execute and assert_receive in Elixir. :)

**Comment by whatyouhide** [comment]
@josevalim added examples 👍 

@arkgil rebased
 🙃

**Review by arkgil — APPROVED** [review]

**Pull request merged**